cylc/flow/network/publisher.py

```diff
@@ -0,0 +1,132 @@
+# THIS FILE IS PART OF THE CYLC SUITE ENGINE.
+# Copyright (C) 2008-2019 NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Publisher for suite runtime API."""
+
+import asyncio
+from threading import Thread
+
+import zmq
+
+from cylc.flow import LOG
+from cylc.flow.exceptions import CylcError
+from cylc.flow import __version__ as CYLC_VERSION
+
+
+async def gather_coros(coro_func, items):
+    """Gather multi-part send coroutines"""
+    try:
+        gathers = ()
+        for item in items:
+            gathers += (coro_func(*item),)
+        await asyncio.gather(*gathers)
+    except Exception as exc:
+        LOG.error('publisher: gather_sends: %s' % str(exc))
+
+
+def serialize_data(data, serializer):
+    """Serialize by specified method."""
+    if callable(serializer):
+        return serializer(data)
+    elif isinstance(serializer, str):
+        return getattr(data, serializer)()
+    return data
+
+
+class WorkflowPublisher:
+    """Initiate the PUB part of a ZMQ PUB-SUB pair.
+
+    This class contains the logic for the ZMQ message Publisher.
+
+    Note: Security TODO
+
+    Usage:
+        * Define ...
+
+    """
+    # TODO: Security will be provided by zmq.auth (post PR #3359)
+
+    def __init__(self):
+        self.port = None
+        self.context = zmq.Context()
```

Comment:
Might want to share the same ZMQ context as `cylc.flow.network.server`.

Reply:
Fixed - I just pass the "server" context to instantiation of the publisher... The network code might need to be re-organised (seems a bit messy as it is?)

Reply:
Re-organise away!

Reply:
Yes - I think I'll wait until Sadie's #3359 is in.. And I don't have a clear idea on how to organise it yet, only that it seems weird that REQ/REP is in a module named `server` & `client` and PUB/SUB in a modules named `publisher` & `subscriber` ..  Perhaps the "server" should setup both, and be more generic? (instead of ZMQServer specialising in REQ/REP)... IDK thoughts?

Reply:
Not so fussed, try to keep the "server" (i.e. workflow process) and "client" logic separate.

Reply:
Done. All 4 classes (REQ/REP, PUB/SUB) inherit `ZMQSocketBase`, which I put in `network/__init__.py` (I couldn't think of a more appropriate place)... 

They just overwrite some class methods where needed..